pkg/v1/v1util/matcher.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1util
+
+import (
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ManifestMatcher function that is given a v1.Descriptor, and returns whether or
+// not it matches a given rule. Can match on anything it wants in the Descriptor.
+type ManifestMatcher func(desc v1.Descriptor) bool
```

Comment:
`ManifestMatcher` as a name is too specific, we can reuse this type to match any descriptor, which can reference arbitrary things. We could just call this a `Matcher`.

If we end up in a package called `match` or `matcher` or whatever, we may want to simply call this thing a `Predicate`.

Reply:
> we can reuse this type to match any descriptor, which can reference arbitrary things

That is a good point. Within `Index`, we are passing it through the descriptors in the manifests, but it can be for anything.

> If we end up in a package called match or matcher or whatever, we may want to simply call this thing a Predicate to avoid stuttering.

Are we moving it to `match` or `matcher` package? I am happy to do it, we just need to decide.

> Predicate

Bit of a mouthful. I am not sure people will understand it at first glance, or know to look for it. It is, well, a "matcher". I know it stutters, but it still is easier to understand than "predicate".

Reply:
I'd be fine with `match.Matcher` as well.

Reply:
😂

Those of us with engineering or mathematical training and/or experience tend to belittle those gut feelings; it is a mistake on our part. Guts count.

In line with what you wrote earlier, this really is not a _manifest_ matcher, but a _descriptor_ matcher. So I put it in `match` package as `Descriptor`, but that feels wrong. It is a Descriptor _matcher_. But then it gets ugly: `match.DescriptorMatcher`? Of course, we might have an `IndexMatcher` and a `ManifestMatcher` and even `BlobMatcher` in the future. Thoughts?